src/_igraph/graphobject.c

```diff
@@ -7742,6 +7742,95 @@ PyObject *igraphmodule_Graph_layout_sugiyama(
   return (PyObject *) result_o;
 }
 
+  /** \ingroup python_interface_graph
+ * \brief Places the vertices of a graph using Uniform Manifold Approximation and Projection (UMAP)
+ * \return the calculated coordinates as a Python list of lists
+ * \sa igraph_layout_umap
+ */
+PyObject *igraphmodule_Graph_layout_umap(igraphmodule_GraphObject * self,
+                                        PyObject * args, PyObject * kwds)
+{
+  static char *kwlist[] =
+    { "dist", "dim", "min_dist", "epochs", "sampling_prob", NULL };
+  igraph_matrix_t m;
+  igraph_vector_t *dist = 0;
+  /* FIXME: use more flexible defaults */
+  double min_dist = 0.01;
+  double sampling_prob = 0.5;
+  Py_ssize_t dim = 2;
+  Py_ssize_t epochs = 500;
+  PyObject *dist_o = Py_None;
+  PyObject *result_o;
+
+  if (!PyArg_ParseTupleAndKeywords(args, kwds, "|Ondnd", kwlist, &dist_o,
+                                   &dim, &min_dist, &epochs, &sampling_prob))
+    return NULL;
+
+  CHECK_SSIZE_T_RANGE_POSITIVE(dim, "number of dimensions");
+  CHECK_SSIZE_T_RANGE_POSITIVE(epochs, "number of epochs");
+
+  if (dist_o != Py_None) {
+    dist = (igraph_vector_t*)malloc(sizeof(igraph_vector_t));
+    if (!dist) {
+      PyErr_NoMemory();
+      return NULL;
+    }
+    if (igraphmodule_PyObject_to_vector_t(dist_o, dist, 0)) {
+      free(dist);
+      return NULL;
+    }
+  }
+
+  if (igraph_matrix_init(&m, 1, 1)) {
+    if (dist) {
+      igraph_vector_destroy(dist); free(dist);
+    }
+    igraphmodule_handle_igraph_error();
+    return NULL;
+  }
+
+  if (dim == 2) {
+    if (igraph_layout_umap(&self->g, dist, &m,
+          (igraph_real_t)min_dist,
+          (igraph_integer_t)epochs,
+          (igraph_real_t)sampling_prob)) {
+      if (dist) {
+        igraph_vector_destroy(dist); free(dist);
+      }
+      igraph_matrix_destroy(&m);
+      igraphmodule_handle_igraph_error();
+      return NULL;
+    }
+  } else if (dim == 3) {
+    if (igraph_layout_umap_3d(&self->g, dist, &m,
+          (igraph_real_t)min_dist,
+          (igraph_integer_t)epochs,
+          (igraph_real_t)sampling_prob)) {
+      if (dist) {
+        igraph_vector_destroy(dist); free(dist);
+      }
+      igraph_matrix_destroy(&m);
+      igraphmodule_handle_igraph_error();
+      return NULL;
+    }
+  } else {
+    /* FIXME: complain about the dimension */
+    if (dist) {
+      igraph_vector_destroy(dist); free(dist);
+    }
+    igraph_matrix_destroy(&m);
+    igraphmodule_handle_igraph_error();
```

Comment:
This won't work here because igraph did not raise any error if we managed to get to this point. You need to do a `PyErr_SetString(PyExc_ValueError, "dim must be 2 or 3")` or something like that.